docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+1. The user runs `uniffi-bindgen generate -l kotlin -o <OUT_PATH> <UDL_PATH>`
+1. `uniffi-bindgen` gets its own version, let's say `0.16` as an example.
+1. `uniffi-bingen` shells out to `uniffi-bingen-kotlin -o <OUT_PATH> -v 0.16 <UDL_PATH>`
+1. `uniffi-bindgen-kotlin` asserts that it was compiled using the same `0.16` version, otherwise it panics.
+1. `uniffi-bindgen-kotlin` then calls a `generate_bindings` function in `uniffi-bindgen` (as a library dependency)
+1. `uniffi-bindgen`'s `generate_bindings` takes a generic parameter type that implements a trait `BindingGenerator`.
+1. `uniffi-bindgen` parses the `udl`, then calls the generic type's `write_bindings` function (the implementation lives in `uniffi-bingen-kotlin`)
+
+
+## **Implementation Option 2: Have `uniffi_bindgen` get dynamically linked with exactly one `uniffi_bindgen_{language}`**
+This approach uses the specific `uniffi_bindgen_{language}` as plugins. It essentially replaces the shelling out in `Implementation Option 1` with calling an exposed unsafe `C` API in a dynamically linked crate. 
+
+I'm not 100% sure how to let `uniffi_bindgen` choose which crate it gets dynamically linked with on run time, but it's possible.
+
+
+## **Implementation Option 3: Have each `uniffi_bindgen_{language}` be run directly**
+This approach has `uniffi_bindgen` only provide scaffolding and traits for each `uniffi_bindgen_{language}` to implement. `uniffi_bindgen` would also include some generic functions that can be called from `uniffi_bindgen_{language}` by treating `uniffi_bindgen` as a library dependency. The flow would look like this:
+
+1. A user would invoke `uniffi-bindgen-kotlin bindings -o <OUT_PATH> <UDL_PATH>`
+1. `uniffi-bindgen-kotlin` would call a generic function `generate_bindings` in `uniffi-bindgen` that takes type parameter `B: BindingGenerator`
+1. `uniffi-bindgen` parses the `udl`, then calls the generic type's `write_bindings` function (which lives in `uniffi-bingen-kotlin`)
+
+Version compatibility is a serious consideration with this approach, and we would benefit from exposing a `scaffolding` option to `uniffi-bindgen-kotlin`. This way the same `uniffi-bindgen-kotlin` is used to generate both scaffolding and bindings. We can also add a warning to `uniffi-bindgen`'s scaffolding that we recommend using a specific binary (like `uniffi-bindgen-kotlin`) to generate the scaffolding as well. Note: The scaffolding generating logic would still live in `uniffi-bindgen` we would only expose it to be called. `uniffi-bindgen-kotlin` would not modify the logic at all, this is purely for version compatibility.
+
+<br />
+<br />
+
+The old prototype in https://github.com/mozilla/uniffi-rs/pull/997 did Implementation Option 1, **I suggest we do Implementation Option 3 as its easier to reason about.**
```

Comment:
Can you say more about the "easier to reason about" piece here? I believe you, because you've tried it 😄.  But I don't have a good sense of what extra cognitive load gets added from Option 1. Is it the version-checking in the handoff to the language-specific binary?

Reply:
Yeah it's pretty much the handoff that made me think this, and the hopping across processes makes it a little harder to debug - for me on the debugger - but I do agree with your note that we can take advantage of the handoff to provide more descriptive error messages about version mismatches, etc.

**Seeing the feedback from you and @badboy makes me want to revert back to option 1!** I think I'm now convinced that
1. The benefit from losing that extra hop out of `uniffi_bindgen` isn't worth having our consumers try to enforce compatibility across different `uniffi_bindgen_{lang}`(especially like Glean and App Services where we have multiple bindings at a time)
2. `uniffi_bindgen` staying as the source of truth (with it passing it's version to the other crates) has a good benefit of reducing metal load to keep track of which binary to call